pkg/clients/service.go

```diff
@@ -0,0 +1,50 @@
+package clients
+
+import (
+	"context"
+	"go-booking-service/commons"
+	"time"
+)
+
+type ClientsService interface {
+	Authorize(context.Context, string, string) (string, error)
+	Validate(context.Context, string) (string, error)
+}
+
+type clientsService struct {
+	encoder EncoderDecoder
+	users   map[string]string
+}
+
+type EncoderDecoder interface {
+	Encode(string, string, time.Time) (string, error)
+	Decode(string, string) (string, error)
+}
+
+func NewClientsServer(e EncoderDecoder, users map[string]string) ClientsService {
```

Comment:
Intenta dejar lo menos dependiente el consumidor de esta función a los internos de esta capa, por ejemplo si el día de mañana quieres cambiar el tipo de dato que vas a usar o inclusive por si vas a agregar mas parametros tendrías que cambiar toda tus invoaciones "consumidores"

revisa esta aproximacion [https://medium.com/@petomalina/dealing-with-optional-parameters-in-go-9780f9bfbd1d](url) Using Variadic Functions with Option Functions

te dejo un ejemplo que escribí para que te quede claro en el momento de implementarlo, revisalo y me cuentas,
[ejemplo](https://play.golang.org/p/a8lWjgofSuM)

Con esto si el día de mañana quieres pasar otro argumento simplemente colocas otro ServiceOption pero to API no se ve afectada, no cambias la firma solo aumentas las capacidades

Reply:
Esa estructura la estoy reemplazando por una conexion a una base de datos. Voy a usar esto para plantearlo de forma de poder cambiar entre distintas bases de datos, con opciones para inicializarlas con valores si es necesario.